feat(discoverv2) - Add a download as CSV button

wmak opened the pull request:
- Added a downloadAsCsv function that's mostly identical to the v1
  version, I wanted to have it separate to not impact v1 at all.
- Using the page title as the csv filename, eg. if you're on project
  summary the export will include that in the filename.
- TODO: need to change the export icon still
![download as csv](https://user-images.githubusercontent.com/4205004/72370972-562bd780-36d1-11ea-997d-7e866e3573a9.gif)

wmak committed:
feat(discoverv2) - Add a download as CSV button

- Added a downloadAsCsv function that's mostly identical to the v1
  version, I wanted to have it separate to not impact v1 at all.
- Using the page title as the csv filename, eg. if you're on project
  summary the export will include that in the filename.
- TODO: need to change the export icon still

markstory commented on src/sentry/static/sentry/app/components/gridEditable/index.tsx:
```diff
@@ -496,7 +498,11 @@ class GridEditable<
           resizing columns. It will be refactored in a upcoming PR */}
           <div style={{display: 'flex', flexDirection: 'row'}}>
-            {this.renderHeaderButton()}
+            <HeaderButton onClick={downloadAsCsv} data-test-id="grid-add-column">
```
```suggestion
            <HeaderButton onClick={downloadAsCsv} data-test-id="grid-download-csv">
```

markstory commented on src/sentry/static/sentry/app/components/gridEditable/index.tsx:
```diff
@@ -496,7 +498,11 @@ class GridEditable<
+              <InlineSvg src="icon-download" />
+              {t('Download CSV')}
+            </HeaderButton>
+            <div style={{marginLeft: '16px'}}>{this.renderHeaderButton()}</div>
```
There are a few marginLeft elements now, can we make a styled component for these? Something like

```typescript
const HeaderButtonContainer = styled('div')`
  margin-left: ${space(2)};
`;
```

markstory commented on src/sentry/static/sentry/app/views/eventsV2/utils.tsx:
```diff
@@ -376,3 +379,26 @@ export function decodeScalar(
+
+export function downloadAsCsv(tableData, filename) {
+  const {meta, data} = tableData;
+  if (meta !== undefined) {
```
You could early return instead.

```typescript
if (meta === undefined) {
  return;
}
```

wmak replied:
:+1: sure thing, but for my learnings what's the difference?

lynnagara replied:
I think we use `typeof(meta) === "undefined"` for checking undefined values

markstory replied:
It saves you a level of indentation.

markstory commented on src/sentry/static/sentry/app/views/eventsV2/utils.tsx:
```diff
@@ -376,3 +379,26 @@ export function decodeScalar(
+    const encodedDataUrl = encodeURI(`data:text/csv;charset=utf8,${csvContent}`);
+
+    // Create a download link then click it, this is so we can get a filename
+    const link = document.createElement('a');
```
The other discover uses `window.location.assign()` did that not work here?

wmak replied:
That works, but you can't change the filename with that method. This way we can include the page title and export time 🕐

lynnagara replied:
neat trick

leedongwei replied:
This trick goes back to jQuery days. Hacky af. I'm disappointed that they (whoever is responsible for web/HTML standards) haven't improved on this after so many years.

wmak committed:
ref: Adding styled button, early returning download function

dashed commented on src/sentry/static/sentry/app/views/eventsV2/utils.tsx:
```diff
@@ -376,3 +379,27 @@ export function decodeScalar(
+
+export function downloadAsCsv(tableData, filename) {
+  const {meta, data} = tableData;
+  if (meta === undefined) {
```
```suggestion
  if (!meta) {
```

dashed commented on src/sentry/static/sentry/app/views/eventsV2/utils.tsx:
```diff
@@ -376,3 +379,27 @@ export function decodeScalar(
   return isString(unwrapped) ? unwrapped : undefined;
 }
+
+export function downloadAsCsv(tableData, filename) {
```
@wmak As discussed, we can pass `eventView` from here https://github.com/getsentry/sentry/blob/eb18bea23be662c9a2c08caedabd410bc535f742/src/sentry/static/sentry/app/views/eventsV2/table/tableView.tsx#L375 to this function.

Then from there we can generate the column order ` const columnOrder = eventView.getColumns();`. This would provide the column order instead of the columns from the `meta`.

You can generate the appropriate keys using this function https://github.com/getsentry/sentry/blob/eb18bea23be662c9a2c08caedabd410bc535f742/src/sentry/static/sentry/app/views/eventsV2/utils.tsx#L247-L256, which would allow you to map the data from `tableData`.

leedongwei reviewed (commented):
Looks pretty reasonable.

wmak committed:
ref: Including column order to keep columns consistent

- With this we can keep the order of columns consistent with the export
  and the UI
- Also the export's headers will match what the frontend shows

wmak committed:
fixup: meta isn't needed for this function anymore

dashed commented on src/sentry/static/sentry/app/components/gridEditable/index.tsx:
```diff
@@ -496,11 +498,17 @@ class GridEditable<
-            {this.renderHeaderButton()}
-
-            <div style={{marginLeft: '16px'}}>
+            <HeaderButtonContainer>
```
@wmak it might be reasonable to hide this button if there is no data. For example, the empty state.

wmak replied:
I did this with `isLoading`, is that what you meant with empty state? Or something else?

dashed replied:
Ah. I had meant when `tableData` was empty.

But, I think this is good enough.

I thought there may be issues with generating the csv data from an empty `tableData`; but it seems to import fine on Google Sheets. 👍

dashed commented on src/sentry/static/sentry/app/views/eventsV2/utils.tsx:
```diff
@@ -376,3 +379,32 @@ export function decodeScalar(
+  const now = new Date();
+  link.setAttribute('href', encodedDataUrl);
+  link.setAttribute('download', `${filename} ${getUtcDateString(now)}.csv`);
+  link.click();
```
@wmak does this work in FireFox and Safari?

Some variations of this appends the link to the DOM before clicking, and removing it afterwards:

- https://github.com/mholt/PapaParse/issues/175#issuecomment-201308792
- https://stackoverflow.com/questions/32225904/programmatical-click-on-a-tag-not-working-in-firefox

wmak replied:
:+1: Works in both, but I'll also add the removing.

wmak committed:
ref: Hide on loading, deleting link after downloading

dashed reviewed (approved):
Looks good to me 👍
